### DISCUSSION TITLE
Equal weight queues for partitioning across redis shards?

### DISCUSSION THREAD

**Discussion by danthegoodman1** [hibiken/asynq#372] [Q&A] [answered]
If I had setup a non-strict weighting where all queues were equal, would this an effective way to increase throughput by partitioning a queue among multiple shards, and then randomly selecting one of these to write the task to?

For example:

```go
srv := asynq.NewServer(redis, asynq.Config{
    Concurrency: 10,
    Queues: map[string]int{
        "partition-1": 1,
        "partition-2": 1,
        "partition-3": 1,
    },
})
```

And then the workers would randomly poll the various queue partitions serving as one larger (virtual) partitioned queue?

Or maybe even for further shuffle sharding effectiveness (and if running for k8s for example) using a StatefulSet where the pod is given a number in the podname (example-1), moduloing that over the known partition count, and then consuming from that queue partition? That way a bad partition wouldn't degrade as many workers (I assume)

Edit: For the single worker with multiple partitions, I see what seems to be an example here https://github.com/hibiken/asynq/wiki/Redis-Cluster

**Comment by hibiken** [comment, marked as answer]
@danthegoodman1 Thank you for opening this discussion, this is a great topic!

You're exactly right. One of the motivations behind adding a support for Redis cluster was so that you can take advantage of automatic  data sharding and increase throughput. 
What you suggested (equal weight with non-strict priority mode) seems correct, you just need to make sure that the partitions hashslots are evenly distributed across the cluster (you may need to reshard to achive that) to fully take advantage of this setup.

Side note:
It may be useful to support "prefix matching" or "regex-matching" in the `Config.Queues` so that you can do something like:

```go
srv := asynq.NewServer(redis, asynq.Config{
    Concurrency: 10,
    Queues: map[string]int{
        "myqueue_partition-*": 1,
    },
})
```

**Reply by danthegoodman1** [reply]
Thanks! Great to hear. I think the prefix matching could be cool especially as the partition count grows.

Using the cli it should be quite easy to make sure that partitions are sharded properly.
